[REP-2001] Add Dev Tools Variant

DLu opened the pull request:
Via https://github.com/ros2/variants/pull/37

Unclear whether I should add it to all the other active variants as well as rolling.

clalancette commented on rep-2001.rst:
```diff
@@ -763,14 +764,24 @@ Simulation
+::
+
+  - dev_tools:
+      packages: [python3-colcon-common-extensions, python3-flake8, python3-pip,
+                 python3-pytest, python3-pytest-cov, python3-rosdep, python3-setuptools,
+                 python3-vcstool, cmake, git, wget]
```
The reason I've been hesitant on this is because this new package is a different kind of beast from all of the other ones in here.  In particular, all of the rest of the variants here depend on things directly generated by the ROS buildfarm (though of course they indirectly depend on things from the underlying operating system).

This new variant, instead, directly depends on a combination of things from the underlying operating system (wget, git, etc), as well as packages that are imported into ROS from elsewhere (python3-colcon-common-extensions).

That doesn't necessarily mean we shouldn't do it, but it is giving me pause both from a precedent standpoint (should we allow other collections of packages?), as well as a practical standpoint (will the buildfarm happily build this?).  I've been talking to @nuclearsandwich and @cottsay a little bit more about this, they may leave their thoughts about this as well.

DLu replied:
(That's not a threat, I more mean that I'm not adamant that it must be part of the REP and needs to go in `variants`)

nuclearsandwich replied:
Cementing some of my thoughts from the discussion Chris alluded to, I think that the biggest outward concern for this metapackage is that by using a ROS package for the variant, we either _only_ release this variant in a single distribution like Rolling or we end up with duplicate variants: `ros-rolling-dev-tools`, `ros-humble-dev-tools`, `ros-iron-dev-tools` which all install the same system-wide packages. This will generate some confusion as users ask questions like "Do I need the dev-tools package for a specific distribution?" or "do I need to source the distribution setup script to use these dev tools?"

Implementing this metapackage as an in-ROS variant is the expedient way to make it happen if those concerns don't worry anyone else. Overall I think having a package like this would greatly improve and streamline the setup instructions for working with source distributions of ROS on supported platforms and in that same discussion we talked about possibly having a `ros-build-essential` package which builds on the platform compiler toolchain to codify and provide the baseline packages which should be present during build farm package builds implicitly.

The rosdistro independent method for creating `ros-build-essential` and `ros-dev-tools` metapackages would be to create them with `equivs` and comparable tools for RPM distributions and distribute them through the ROS bootstrap repositories.

DLu replied:
I get the concern of having duplicate variants, but I worry about asserting that the dependencies will be the same for all distributions forever and always. What if the ghost of Dirk Thomas comes back and writes `colcon2` for ROS-2-K? If you install `ros-build-essential` does it have `python3-colcon-common-extensions` or `python3-colcon2`?

DLu replied:
Is `equivs` a specific repo like `variants`?

nuclearsandwich replied:
> I get the concern of having duplicate variants, but I worry about asserting that the dependencies will be the same for all distributions forever and always.

Yeah, that has crossed my mind as well but as it stands all of the tools in the variant now are ROS independent. For a time ROS 2's ament_package and ament_tools were included in each distribution but both were eventually dropped in favor of independent tools (ament_package is technically still in Rolling but it contains _only_ ROS 2 specific templates and no actual code).

It does not make sense, in my opinion, for this metapackage to include build tools (such as cmake, catkin, ament_cmake) since each package is expected to explicitly depend on the build tool that it uses, but if build tools were included, that would add weight to the ROS-dependent approach.

> What if the ghost of Dirk Thomas comes back and writes `colcon2` for ROS-2-K? If you install `ros-build-essential` does it have `python3-colcon-common-extensions` or `python3-colcon2`?

In point of fact, I think `ros-build-essential` has neither because neither is critical to build a ROS package on the build farm. But `ros-dev-tools` would either include both tools or include only the one that is the recommended build tool for all current distributions as the current proposal does by not including catkin-tools. For bootstrapping reasons, I don't foresee a situation where the officially recommended build tool is ever a ROS-dependent package like ament_tools was again.

> Is `equivs` a specific repo like `variants`?

Whoops, sorry for the jargon. [`equivs`](https://manpages.debian.org/testing/equivs/equivs-build.1.en.html) is a tool for building small Debian packages using just a control file instead of an entire package and `debian/` which makes creating small metapackages really straightforward (assuming you're only supporting Debian-derived platforms). It has been used in ROS 2 to create [trivial packages that satisfy package dependencies](https://github.com/nuclearsandwich/ros2-equivs) on platforms where they are not available since the package.xml format has no way to describe optional/opportunistic dependencies.

DLu replied:
NB: I'm explaining my rationale, but I'll be happy with whatever solution you and other experts come up with. 

All I want at the end of the day is a workflow where I can do this.

    sudo apt install ros-$ROSDISTRO-ros-base
    sudo apt install ros-SOMETHING
    rosdep install --from-paths src --ignore-src -y
    colcon build

nuclearsandwich replied:
> All I want at the end of the day is a workflow where I can do this.

Based on the upbeat tempo of the offline discussion we were having I think you have @clalancette, @cottsay, and myself all convinced that this would be a big improvement.

I think that we've passed the phase where we're resistant to the proposal (which for me was based on the novelty and not some principled opposition) and into the phase where we're trying to think through the implementation options and the effects of each of them to arrive at an optimal long term solution.

nuclearsandwich commented on rep-2001.rst:
```diff
@@ -763,14 +764,24 @@ Simulation
+::
+
+  - dev_tools:
+      packages: [python3-colcon-common-extensions, python3-flake8, python3-pip,
```
I'm curious about the rationale for including flake8 (and to a lesser degree pytest and pytest-cov) when these aren't ubiquitous dependencies for a ROS developer workflow.

DLu replied:
In short, they're on the list here: https://docs.ros.org/en/humble/Installation/Alternatives/Ubuntu-Development-Setup.html#install-development-tools-and-ros-tools

Personally, I would just go with `build-essential cmake git python3-colcon-common-extensions python3-pip python3-rosdep python3-setuptools python3-vcstool` but I was deferring to whoever wrote the installation page.

clalancette replied:
> I'm curious about the rationale for including flake8 (and to a lesser degree pytest and pytest-cov) when these aren't ubiquitous dependencies for a ROS developer workflow.

I'd honestly go even further than this, and add in the packages corresponding to the flake8 plugins that we expect to be there.  We've had lots and lots of problems where people have trouble getting the right combination of flake8 plugins to match CI, and get frustrated by this.  Adding them here won't totally solve it (if people are using pip installed packages, or venvs they could still have the problem), but it will at least help.

nuclearsandwich commented:
Bringing a few more discussions back out into the open: 

I'm in favor of creating two layers of metapackages a `build-essential` or `base-devel` (the latter name coming from the Archlinux community) which includes the set of packages which will be installed whenever building ROS packages on the build farm without having to declare them as explicit dependencies. This would be limited to the host operating system's compiler and packaging toolchain and possibly just python3-catkin-pkg although @cottsay probably knows better what's in that set. As well as the full `dev-tools` metapackage which adds to the essentials the common tools used by most developers and maintainers like `python3-rosdep`, `python3-bloom`, `python3-vcstool`, and `python3-colcon-common-extensions` (perhaps exploded into a set of ROS-specific colcon extensions rather than using that metapackage).

If we create these as ROS distro independent packages, i.e. `ros-dev-tools` it will mark the first time that we've created `ros-*` packages which do not match `ros-$ROSDISTRO-*` this will foil some operations in the build farm stack which include filtrations which import packages per-rosdistro by excluding `ros-*` packages unless they match `ros-$TARGETDISTRO-*` in order to sync all independent packages but exclude packages from other rosdistros from the sync. So we'd have to special case or work around that for independent packages which are still `ros-` prefixed.

DLu commented:
In case you need it, you have my permission to edit this PR or open another alternative PR.

DLu commented:
Hope everyone had a good summer vacation :-D

What's the best move here? @nuclearsandwich Do you want to put together a REP for the two metapackage path? 

Do we need to informally vote on whether it should be distro specific?

nuclearsandwich commented:
Thanks for giving this one a little push. 
Yeah I think there's pretty unanimous agreement that some kind of devtools metapackage would be a useful addition we just need to settle on a method with the various trade-offs and benefits.

> Do we need to informally vote on whether it should be distro specific?

Give this comment a 👍🏼 in favor of a distribution specific metapackage (this is only chosen as the positive because it's the current state of this proposal) or a 👎🏼 in favor of a distribution independent solution.

We'll "close polls" in 48 hours. For the moment I'm interested in the attitudes of those already part of the discussion rather than making a formal poll on Discourse.

DLu commented:
I voted :+1: because it seems like the easiest path forward implementation-wise and it leaves the door open for distro-dependent  build tools in the future. 

I'd be willing to change to :-1: if Open Robotics would commit to the development time required to ensure that the build farm architecture supports distro independent aliases/groups.

nuclearsandwich commented:
Welp. Thread participants have voted and it's a dead heat unless I can flip @DLu 😁 

> I'd be willing to change to -1 if Open Robotics would commit to the development time required to ensure that the build farm architecture supports distro independent aliases/groups.

I think that an independent metapackage would be handled similarly to other ROS infrastructure packages and imported to the buildfarm and ROS repositories via the ROS bootstrap repository.

To that end support already exists although it probably serves as motivation to shore up the automation for producing such metapackages since I currently build them locally and push them directly.

And as discussed above we'd potentially have to update the package import scripts depending on the naming conventions but I think we can bikeshed those details once we know we need new bikes.

Is there specific support beyond that which you would expect?

---

@clalancette assuming the above discussion succeeds, you're the remaining vote in favor of per-distro dev tools packages. Would you be willing to elaborate on your preference? Especially any concerns you have with a rosdistro independent dev tools package.

clalancette commented:
> @clalancette assuming the above discussion succeeds, you're the remaining vote in favor of per-distro dev tools packages. Would you be willing to elaborate on your preference? Especially any concerns you have with a rosdistro independent dev tools package.

The way I came to a +1 on a ROS distribution-specific metapackage is that I went back and reread all of the comments we made on this issue.  And at the end of the day, I found DLu's argument that the development packages may change release-to-release compelling.  While it is absolutely true that our tools are pretty stable (and have been at least back through Dashing), it is not necessarily the case that this won't change in a future ROS distribution.  So it seems to me that the most future-compatible way to do this is with a ROS distribution-specific metapackage.  It also seems like this would be the "easiest" way to get it done.

All of that said, I don't really feel strongly about it.  If you still think that a ROS distribution-agnostic package is the way to go, I'd be happy enough to have a `ros-dev-tools` package rather than `ros-humble-dev-tools`.  If changes come in later ROS distributions, we'll just have to find another way to deal with it.

nuclearsandwich commented:
Thanks clalancette for that info. I'm adding this to the Open Robotics Infrastructure team's discussion topics when we meet on Monday and we'll talk through specifics of an independent implementation, then we'll bring those specifics back here (as well as to the ROS core team meeting the following day).

nuclearsandwich commented:
@cottsay and I talked it through and both of us think that the likelihood of _incompatible_ differences between build tools for different ROS 2 distributions is infinitesimally small. The confusion potential for explaining that in order to build for example, ROS Rolling from source starts with installing `ros-rolling-dev-tools` from apt as well as around whether you need a dev-tools package for each distribution your working with is great enough that using the existing ROS variants repo or a separate variant-style metapackage within ROS isn't what we'd like to do.

Instead we'd like to create and maintain these packages as the ROS infrastructure team using platform-specific methods: a bare control file and equivs-build on Debian-derived distros and a bare spec file for rpm-using distros and publish them via the ROS bootstrap repository from which they'll be imported to the ROS build farm and then the ROS and ROS 2 repositories as well as any other build farm out in the community.

@cottsay also agreed with my idea for creating separate "build essential" and "developer tools" metapackages which distinguish between the essentials you need to build ROS and the common set of user-friendly frontend tools which are not, strictly speaking, essential to building ROS packages. Although his main motivation is to prevent singleton handling from creeping in and making sure that we always maintain a plurality of these metapackages rather than one special one.

I had some concerns about what a package name starting with `ros-` but independent of any rosdistro would do to the infra but the entire build farm handles it just fine. It's only the snapshot repo scripts which need to be updated to handle these new packages and that's an easy change to make and deploy.

Furthermore, we think that we can schedule the work on this for October (after some other in-progress infra work has settled down) so we can _potentially_ have the metapackages available in time for ROSCon in Kyoto.

With that in mind, the remaining questions are:

1. What do we actually call these things.
  For the packages absolutely required to build ROS I've put forward `ros-build-essential` and `ros-base-devel`, the former matching Debian's `build-essential` and the latter from Arch Linux which is slightly shorter.
  For the developer friendly tools on top of that `ros-dev-tools` would be the independent package name, are we worried. Does spelling out `ros-developer-tools` or `ros-development-tools` provide any advantage?
  
2. Should we continue to specify these by updating REP 2001, create a new REP to specifically discuss these infrastructural metapackages, or call them an "implementation detail" and not specify them at all! (I am not in favor of this final option)?

3. Just to stir the pot as I went back to read the original [hit tweet](https://twitter.com/TheRealFergs/status/1531439243985313793) should we update the desktop variant to include the dev tools metapackage once it is ready?

clalancette commented:
> Instead we'd like to create and maintain these packages as the ROS infrastructure team using platform-specific methods: a bare control file and equivs-build on Debian-derived distros and a bare spec file for rpm-using distros and publish them via the ROS bootstrap repository from which they'll be imported to the ROS build farm and then the ROS and ROS 2 repositories as well as any other build farm out in the community.

:+1:

> 1. What do we actually call these things.
>        For the packages absolutely required to build ROS I've put forward `ros-build-essential` and `ros-base-devel`, the former matching Debian's `build-essential` and the latter from Arch Linux which is slightly shorter.
>        For the developer friendly tools on top of that `ros-dev-tools` would be the independent package name, are we worried. Does spelling out `ros-developer-tools` or `ros-development-tools` provide any advantage?

I like `ros-build-essential`, as it mirrors Debian.  I also prefer `ros-dev-tools`, as it is pretty clear what it is for.  But I don't feel strongly about the name for either one.

> * Should we continue to specify these by updating REP 2001, create a new REP to specifically discuss these infrastructural metapackages, or call them an "implementation detail" and not specify them at all! (I am not in favor of this final option)?
> 
> *

I'll quote from REP-2001:

```
This document covers the common variants used for recommended installations. It is expressed in the ros2/variants repository on github.com [[2]](https://www.ros.org/reps/rep-2001.html#id32).
```

In short, if we are going to maintain the metadata for the new development packages in https://github.com/ros2/variants, then I'll suggest we add it to REP-2001.  If we are going to maintain the metadata elsewhere, then I'd be for having a new REP.

> 3\. Just to stir the pot as I went back to read the original [hit tweet](https://twitter.com/TheRealFergs/status/1531439243985313793) should we update the desktop variant to include the dev tools metapackage once it is ready?

After thinking about it a bit, yes, I think the `desktop` variant should depend on it.  But my conviction on this one is really weak, so if others think otherwise I'll easily cave on this.

DLu commented:
> In short, if we are going to maintain the metadata for the new development packages in https://github.com/ros2/variants, then I'll suggest we add it to REP-2001. If we are going to maintain the metadata elsewhere, then I'd be for having a new REP.

I will update this PR with a working proposal if @nuclearsandwich can answer the question about the implementation.

nuclearsandwich commented:
> > In short, if we are going to maintain the metadata for the new development packages in https://github.com/ros2/variants, then I'll suggest we add it to REP-2001. If we are going to maintain the metadata elsewhere, then I'd be for having a new REP.

> I will update this PR with a working proposal if @nuclearsandwich can answer the question about the implementation.

In short, no we'll use a new repository under the [ros-infrastructure](https://github.com/ros-infrastructure) org to implement these. But there are further thoughts on REP organization below.

Notes from today's ROS 2 core team meeting (with gratitude to @clalancette for serving as scribe):

### On the subject of splitting out of this REP

We're going to try and add it to REP-2001 and if the wording is odd or if it feels tacked-on we'll move it to a separate 2k rep.

### On the subject of independent metapackages

> If we make the dev tools metapackage distro-independent, then it can’t evolve as quickly.

This is true and one of the major reasons why ROS 2's earliest build tool: ament_tools, was built and bootstrapped in-workspace with ROS 2. It is still possible to evolve over time things by adding new tools alongisde the existing ones then eventually remove the older tools once all actively supported distributions have transitioned away from them. Since all of the dev tools are currently distro independent, it makes the most sense for a metapackage collecting them to be independent as well.

> @sloretz brought up the fact that ROS 1 had to weather a transition from Python 2 to Python 3 where Melodic and Foxy were both on the same platform but used different versions of Python internally and how would an independent dev tools package handle this.

This is actually a sound point and probably the strongest case I've yet seen for distro-specific dev tools packages. However, I think that the versions of dependencies used is primarily a _platform_ question not a distribution question and that for any of the platform dependencies which undergo an epochal shift (such as theoretical incompatible python>3 or cmake>3) would be handled at the platform level rather than the distribution level.

### On the subject of including the dev tools metapackage in desktop

> Does `desktop` as it stands imply a "developer" desktop or a "user" desktop? Are there users who aren't developers in the sense of building packages? Python-only users may just write scripts and never write a package Python or otherwise.

> Desktop is already quite the kitchen sink and adding the devtools won't significantly impact its size. The question was asked if gcc / the compiler toolchain already gets pulled in by desktop as that set of transitive dependencies could significantly inflate the size of a desktop system.

gcc is already present on a desktop docker image. I'll do some explicit tests to determine what the transitive dependency size impact is when including dev packages.

### On the subject of build-essential

> Do we need a build-essential metapackage separate from user-facing dev tools?

The infrastructure team wants one as it would help us resolve platform-specific differences in development packages across platforms. It doesn't have to appear in any user- or developer-facing documentation outside of infrastructure development.

---

We decided, rather than let the discussion run even longer with names to have the discussion of them on this thread and slug it out here. Let's do this [King of the Hill](https://en.wikipedia.org/wiki/King_of_the_Hill_(game)) style.

The current names are:

* `ros-build-essential` for the infrastructure team's packages strictly required to build ROS packages.
* `ros-dev-tools` for the package which includes all the tools a user would want if they're developing ROS packages.

If you think a package should be named differently, propose another name and try to push these off the hill.

nuclearsandwich commented:
I'm assuming that [this](https://github.com/ros-infrastructure/rep/pull/355#issuecomment-1245827968) is a jest referencing the name of my name selection method (and a good one!) rather than a serious suggestion in which case `ros-build-essential` and `ros-dev-tools` stand unchallenged.

DLu commented:
I have updated the PR to now reference `build_essential` and `dev_tools` and to be in a separate section for Distro Independent. Open questions from my end:
 * There is an explicit reference in the REP to the ros2/variants repo. Would probably make sense to put the [`ros-infrastructre`](https://github.com/ros-infrastructure) repo in there too. 
 * There's not a ton of discussion of what variants mean in the context of ROS distros, so there's probably a section missing on saying that some end-user entry points are distro specific and (now) some aren't. 
 * The REP seems to name things with underscores (`desktop_full`) rather than hyphens (`python3-colocon-common-extensions`), so I changed the agreed upon names, i.e. `build-essential` => `build_essential`. I am happy to conform to whichever style. 
 * The infrastructure team should review my choices for what was "essential".

clalancette commented on rep-2001.rst:
```diff
@@ -773,6 +775,33 @@ Desktop Full
+ROS Distribution Independent (September 2022 - Ongoing)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+These variants are not released as ROS packages within each distribution as the others are.
+Instead they are made available as independent packages outside of any ROS distribution.
```
I have two comments here which are related.

First of all, one outcome I would love is from this work is for us to be able to simplify the instructions in https://docs.ros.org/en/rolling/Installation/Alternatives/Ubuntu-Development-Setup.html .  In particular, I would like it if https://docs.ros.org/en/rolling/Installation/Alternatives/Ubuntu-Development-Setup.html#install-development-tools-and-ros-tools was reduced to `sudo apt-update && sudo apt install -y ros-dev-tools`.

To get there, we would need to add all of the development tools from that documentation page to the `ros-dev-tools` variant.  Do we want to do that here?  Alternatively, we could add most of those tools as `package.xml` dependencies...somewhere, though it isn't totally clear to me where.

Depending on the answer to that question, we may have to scope the `ros-dev-tools` to be Ubuntu 22.04 and forwards.  That's because many of the packages in that documentation list were not available until Jammy.

nuclearsandwich replied:
I get into this in a separate line comment https://github.com/ros-infrastructure/rep/pull/355#discussion_r976894371

Going into each of the packages in that section.


* build-essential: We have to hash out some cross-platform specifics but this is in ros-dev-tools via ros-build-essential.
* cmake: in ros-dev-tools via ros-build-essential
* git: in ros-dev-tools via ros-build-essential (git is a requirement of several vendor packages but we also encourage explicit declaration where it is used so perhaps this could move to dev-tools directly)
* python3-colcon-common-extensions: ros-dev-tools
* flake8 + plugins: We should declare these in ament_flake8, it's been on my wishlist ever since we got them packaged in Ubuntu it just hasn't been my top priority:
  * python3-flake8
  * python3-flake8-blind-except
  * python3-flake8-builtins
  * python3-flake8-class-newline
  * python3-flake8-comprehensions
  * python3-flake8-deprecated
  * python3-flake8-docstrings
  * python3-flake8-import-order
  * python3-flake8-quotes
* python3-pip: I think we can actually drop this entirely, it used to be for installing flake8 extensions from pip but we install everything via apt now. "pip install" does not appear anywhere in that doc. 
* pytest and plugins: These are dependencies of [colcon-core](https://github.com/colcon/colcon-core/blob/0bf904e205aaae3055f9b8fefbdd4f52806b66e6/setup.cfg#L33-L39) although it's noted that they are technically conditional. I think their conditionality may have been due to the fact that they were not all available via apt. But either way colcon requires them to run tests. I could see these being "build essential" by virtue of being "test essential" but it sort of comes down to taste and convention. We expect every package to declare a dependency on its _build_ framework (`cmake` for `cmake`, `ament_cmake` for `ament_cmake`, etc) why not make them declare their test framework as well. The `ament_python` build type used to have a corresponding library but issues with how dependencies are specified in python and what is available during a sourcepkg job caused us to remove that (although new developments have made it potentially possible again in the future). I would expect such a build type to require pytest but as that's a hypothetical I'm not sure what the best thing to do today is yet. It irks me to think that these packages would be required here simply because they don't fit anywhere else.
  * python3-pytest
  * python3-pytest-cov
  * python3-pytest-repeat
  * python3-pytest-rerunfailures
* python3-rosdep: ros-dev-tools
* python3-setuptools: ros-build-essential
* python3-vcstool: ros-dev-tools
* wget: hah, this one I thought I actually tried to remove in favor of curl. It's in the instructions in order to fetch the ros2.repos file from GitHub. I am not sure that swapping wget for curl makes me feel any better but it is what we should do.

clalancette replied:
One overall comment that I just realized; all of these package names should be `rosdep` names, not Ubuntu package names.  Otherwise we can never make this work for other operating systems.  With that said, here are some comments on the package list (if I agree with your statement, I've just not said anything for brevity):

> build-essential: We have to hash out some cross-platform specifics but this is in ros-dev-tools via ros-build-essential.

But it is currently not explicitly listed in `ros-build-essential`.  Also, there is no rosdep key for it.

> flake8 + plugins: We should declare these in ament_flake8, it's been on my wishlist ever since we got them packaged in Ubuntu it just hasn't been my top priority

All right, that is fair enough.

>  python3-pip: I think we can actually drop this entirely, it used to be for installing flake8 extensions from pip but we install everything via apt now. "pip install" does not appear anywhere in that doc.

Good point.  I'm fine with dropping it.

> pytest and plugins: ... why not make them declare their test framework as well

I'm not sure I would classify the use of those plugins as a "test framework".  Pytest is the test framework, and most (if not all) packages that use it have an explicit dependency in their `package.xml` on the pytest rosdep key.  Some of them also declare dependencies on other `pytest` extensions, though that is not universal.  The issue here is that typically the package itself doesn't depend on `rerunfailures` and such; those are requirements imposed upon it by the external tool (usually `colcon`).  So this is a long way of saying that I think those plugins should be dependencies of `colcon-core` (which they currently are), and we should not add them here. (incidentally, `python3-pytest-cov` is a `Recommend` for `colcon-core`, not a straight-up dependency.  Maybe we should change that)

> wget: hah, this one I thought I actually tried to remove in favor of curl. It's in the instructions in order to fetch the ros2.repos file from GitHub. I am not sure that swapping wget for curl makes me feel any better but it is what we should do.

Actually, `wget` isn't used for the default instructions anymore; we switched to having `vcstool` directly get the file over HTTP (since it can do that).  That said, we use `wget` in enough places in the documentation (I count 47 uses) that I think we should add it here, even though it largely duplicates functionality with curl.

nuclearsandwich replied:
> Actually, `wget` isn't used for the default instructions anymore; we switched to having `vcstool` directly get the file over HTTP (since it can do that). That said, we use `wget` in enough places in the documentation (I count 47 uses) that I think we should add it here, even though it largely duplicates functionality with curl.

😖

> One overall comment that I just realized; all of these package names should be `rosdep` names, not Ubuntu package names.

The rosdep names for all but the most venerable of keys should match the Ubuntu package names so this is probably not a significant issue.

clalancette replied:
Sounds good.  This generally looks good to me now.

clalancette commented on rep-2001.rst:
```diff
@@ -773,6 +775,33 @@ Desktop Full
+::
+
+ - ros-build-essential:
+      packages: [cmake, git, python3-colcon-common-extensions, python3-rosdep,
+                 python3-setuptools, wget]
```
Two comments on this variant.  First, do we want to add `build-essential` to this list?  I think we discussed this before, but I don't remember the outcome of that discussion.

Second, do we really want `python3-colcon-common-extensions` in `ros-build-essential`, or should we move it into `ros-dev-tools`?  I ask because my understanding is that we wanted `ros-build-essential` to be available for the buildfarm to depend on.  But the binary and source packaging jobs on the buildfarm do not strictly depend on colcon, so it is unnecessary dependency there.  On the other hand, PR and dev jobs probably *do* depend on colcon, so maybe we just want to take the hit on that.  Just food for thought; I don't feel strongly either way.

nuclearsandwich replied:
```suggestion
 - ros-build-essential:
      packages: [cmake, git, python3-setuptools, wget]
```

I think these are the critical packages with the possible addition of python3 itself to support `ros_buildfarm` scripts. I'd be inclined to add `python3-ros-buildfarm` if we were using the release version but the official farm tracks HEAD.

`rosdep` and `colcon` aren't used to actually build binary packages on the build farm since by the time those builds are run on individual packages after bloom has already resolved rosdep keys to platform packages.

nuclearsandwich replied:
```suggestion
 - ros-build-essential:
      packages: [cmake, git, python3-setuptools, wget]
```

I think these are the critical packages with the possible addition of python3 itself to support `ros_buildfarm` scripts. I'd be inclined to add `python3-ros-buildfarm` if we were using the release version but the official farm tracks HEAD.

`rosdep` and `colcon` aren't used to actually build binary packages on the build farm since by the time those builds are run on individual packages after bloom has already resolved rosdep keys to platform packages.

nuclearsandwich commented:
> I have updated the PR to now reference `build_essential` and `dev_tools` and to be in a separate section for Distro Independent. Open questions from my end:

Hah, I saw your recent changes when I fetched the branch locally but didn't come and read your questions. I addressed a few of them unknowingly in some just pushed contributions.




> * There is an explicit reference in the REP to the ros2/variants repo. Would probably make sense to put the [`ros-infrastructre`](https://github.com/ros-infrastructure) repo in there too.

Yeah, we did a bit of bikeshedding on the infra team and I put the to-be-created repository in https://github.com/ros-infrastructure/rep/pull/355/commits/3b5b238ac73d40de4e6206b4ecfe31be4e09b646


> There's not a ton of discussion of what variants mean in the context of ROS distros, so there's probably a section missing on saying that some end-user entry points are distro specific and (now) some aren't.


I had started by splitting the new independent entry points out when I started to wonder if maybe that section isn't the place to go into it. In [b97152e](https://github.com/ros-infrastructure/rep/pull/355/commits/b97152ea58276a807e513b00d547a05e10b18831) I also specified that those two entry points are ROS distribution independent but I did not elaborate on what that means expecting it to be described below.

I haven't pushed this change yet but I'm also inclined to propose dropping `ros-build-essential` from the user facing entry points as I think that nearly everyone should be pushed to `ros-dev-tools` instead with the primary consumer of the `ros-build-essential` package being the build farm.

> * The REP seems to name things with underscores (`desktop_full`) rather than hyphens (`python3-colocon-common-extensions`), so I changed the agreed upon names, i.e. `build-essential` => `build_essential`. I am happy to conform to whichever style.

I hadn't seen this but in [b97152e](https://github.com/ros-infrastructure/rep/pull/355/commits/b97152ea58276a807e513b00d547a05e10b18831) I moved the independent variants to their "platform package names" with the rationale being that unlike the other variants, these are not package.xml manifest package names which will get "rendered" as in-distro names (i.e. `ros_core` -> `ros-rolling-ros-core`) I think that the names we use here should match to what we'd expect to put in package.xml dependency keys: For packages in a ROS distribution that's the name of the package in package.xml but for system packages, which is essentially what the independent variants will be, the rosdep key name defaults to the name of the package in Debian format. I also added the `ros-` to the names everywhere along the same rationale.

> * The infrastructure team should review my choices for what was "essential".

Yeah, I haven't pushed any commits on this because I also want to get feedback so I'm going to do those as "suggestions" using inline code review. I do need to continue points from [an earlier](https://github.com/ros-infrastructure/rep/pull/355#discussion_r910227721) discussion on that front as well.

Thanks for keeping the ball rolling on this one!

clalancette reviewed (commented):
I've left a few more comments/questions.  Overall, I feel like this is coming together.

nuclearsandwich commented on rep-2001.rst:
```diff
@@ -773,6 +775,33 @@ Desktop Full
+
+::
+
+  - ros-dev-tools:
+      extends:  [ros-build-essential]
+      packages: [python3-flake8, python3-pip, python3-pytest,
+                 python3-pytest-cov, python3-vcstool]
```
```suggestion
  - ros-dev-tools:
      extends:  [ros-build-essential]
      packages: [python3-bloom, python3-colcon-common-extensions,
                 python3-rosdep, python3-vcstool]
```

We discussed this on an earlier revision https://github.com/ros-infrastructure/rep/pull/355#discussion_r908982708 but I don't think the pytest or flake8 packages belong here. I think that those should be per-package dependencies and moved out of the documentation where they occur.

The original reason they ended up there was because not all of the flake8 or pytest extensions used were available as apt packages in ROS 2's target platforms. They're available as-of Jammy and so we'll have to decide if it's worth adding them in dev-tools for the sake of ROS 2 Foxy.

I should have considered this as an explicit case for how to handle differences between distros for independent metapackages but I didn't because I don't think they belong in this metapackage for any distro in the first place.

DLu commented:
I was going to make a separate issue for the `python3-flake8` dependencies, but it already exists here: https://github.com/ament/ament_lint/issues/254

DLu commented:
Hey, so what's the next step here? Did we come to enough consensus that I should edit the proposal?

nuclearsandwich commented on rep-2001.rst:
```diff
@@ -773,6 +775,33 @@ Desktop Full
+
+ - ros-build-essential:
+ - ros-build-essential:
+      packages: [cmake, git, python3-setuptools, wget]
```
`wget` isn't needed for package builds and should be in ros-dev-tools instead.

```suggestion
      packages: [cmake, git, python3-setuptools]
```

nuclearsandwich replied:
@cottsay should we make Python 3 itself explicitly required?

```suggestion
      packages: [cmake, git, python3, python3-setuptools]
```

cottsay replied:
> should we make Python 3 itself explicitly required?

It wouldn't hurt to call it out explicitly, especially in an increasingly possible setuptools-free future.

nuclearsandwich replied:
Add python3: [8a2e8ae](https://github.com/ros-infrastructure/rep/pull/355/commits/8a2e8ae56c9a1bc411d1bd341cae67497934a324)

Move wget: [91f54b3](https://github.com/ros-infrastructure/rep/pull/355/commits/91f54b31cd23e368829914d4cbc9d6a696f4f0b5)

nuclearsandwich commented on rep-2001.rst:
```diff
@@ -773,6 +775,33 @@ Desktop Full
+
+  - ros-dev-tools:
+      extends:  [ros-build-essential]
+      packages: [python3-bloom, python3-colcon-common-extensions,
+                 python3-rosdep, python3-vcstool]
```
Move `wget` to the dev-tools variant instead.

```suggestion
      packages: [python3-bloom, python3-colcon-common-extensions,
                 python3-rosdep, python3-vcstool, wget]
```

nuclearsandwich replied:
[91f54b3](https://github.com/ros-infrastructure/rep/pull/355/commits/91f54b31cd23e368829914d4cbc9d6a696f4f0b5)

nuclearsandwich reviewed (commented):
I think this is just about ship-shape.

Just need to reorganize a few last packages.

nuclearsandwich commented on rep-2001.rst:
```diff
@@ -773,6 +775,34 @@ Desktop Full
+Build Essential
+"""""""""""""""
+
+In addition to the explicitly listed packages this variant should include the base toolchain for any operating system (such as `build-essential`) on Debian as well as GNU Make and a C++ compiler if not already part of that toolchain.
```
I don't think there is a way to include this in the extends/packages format below so I pulled the descriptive text from the implementation README in infra-variants. Extra eyes here are appreciated.

DLu replied:
I'm not looking at any of the implementation at the moment, but it seems like there should be a [`build-essential` entry in the `rosdep/base.yaml`](https://github.com/ros/rosdistro/blob/master/rosdep/base.yaml)

nuclearsandwich replied:
> I'm not looking at any of the implementation at the moment, but it seems like there should be a [build-essential entry in the rosdep/base.yaml](https://github.com/ros/rosdistro/blob/master/rosdep/base.yaml)

I'm a bit conflicted about it at the moment. Since the infra-variants aren't ROS packages and don't use rosdep, it's not required for those packages to work and for Debian/Ubuntu specifically, I don't think anything is necessarily _supposed_ to depend on `build-essential` since it is meant to be assumed available when building any package and if a package has dependencies on specific contents of build-essential it should declare those instead.

The reason we're including it for ros-build-essential is that we intend a superset of it.

Since it's not required to have a rosdep key for it, it is not a blocker.

nuclearsandwich reviewed (approved):
With the caveat that I've now pushed my latest preferences (after a confirming review from Scott) I think this is good to go!

The implementation in https://github.com/ros-infrastructure/infra-variants/pull/1 is also coming together.

DLu commented on rep-2001.rst:
```diff
@@ -1,6 +1,6 @@
 REP: 2001
 Title: ROS 2 Variants
-Author: Tully Foote <tfoote@openrobotics.org>, Mikael Arguedas <mikael@openrobotics.org>, Ruffin White <roxfoxpox@gmail.com>
+Author: Tully Foote <tfoote@openrobotics.org>, Mikael Arguedas <mikael@openrobotics.org>, Ruffin White <roxfoxpox@gmail.com>, David V. Lu!! <davidvlu@gmail.com>
```
Y'all should add your names here.

nuclearsandwich replied:
Done in [70575c9](https://github.com/ros-infrastructure/rep/pull/355/commits/70575c978b0162ad5ca1b50c4c27481fbd75df96)

clalancette commented on rep-2001.rst:
```diff
@@ -773,6 +775,34 @@ Desktop Full
+Build Essential
+"""""""""""""""
+
+In addition to the explicitly listed packages this variant should include the base toolchain for any operating system (such as `build-essential`) on Debian as well as GNU Make and a C++ compiler if not already part of that toolchain.
```
```suggestion
In addition to the explicitly listed packages this variant should include the base toolchain for any operating system (such as `build-essential` on Debian) as well as GNU Make and a C++ compiler if not already part of that toolchain.
```

Besides my edits, I'm wondering if we should remove the `GNU Make` reference; someday in the future when we make a Windows version of that, it won't be accurate.  On the other hand, I'm OK with deferring this until we get to that point.

nuclearsandwich replied:
> I'm wondering if we should remove the GNU Make reference; someday in the future when we make a Windows version of that, it won't be accurate. On the other hand, I'm OK with deferring this until we get to that point.

I think it would make more sense to make Windows the special case (since the MSVC toolchain would be the requirement instead). I've had issues building CMake projects in the default ROS configurations when BSD make is the default instead of GNU make.

clalancette commented on rep-2001.rst:
```diff
@@ -45,6 +45,8 @@ We define various entry points for ROS users.
  * `perception`: The `perception` variant includes packages commonly used for perception capabilities.
  * `simulation`: The `simulation` variant includes packages that are needed for automated simulation,
    without extra debugging tools.
+ * `ros-build-essential`: The `ros-build-essential` ROS distribution independent variant includes only the packages that are required for building and compiling ROS 2 packages.
```
I think the problem with this statement is that a user who is unfamiliar with the distinction might think this is enough to compile locally.  So I might reword this to use the wording we are talking about elsewhere, something like:

```suggestion
 * `ros-build-essential`: The `ros-build-essential` ROS distribution independent variant includes only the packages that are required for building ROS 2 packages on the buildfarm.
```

nuclearsandwich replied:
I dislike "on the buildfarm" because it's not entirely correct. It _is_ possible to build ROS packages with just these, but it isn't easy and there's additional effort involved but it's probably more succinct to just equate the build farm with that situation.

However I cannot bring myself to merge an incomplete truth so here's a counter-suggestion:

```suggestion
 * `ros-build-essential`: The `ros-build-essential` ROS distribution independent variant includes only the packages that are required for building standalone ROS 2 packages. ROS developers should use `ros-dev-tools` instead for the full set of recommended tools.
```

clalancette replied:
Your update works for me, thanks.

clalancette reviewed (commented):
I left a few comments, but they are really all minor wording updates.  Overall looks good to me.

nuclearsandwich commented:
💯 thanks @DLu for championing this!

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-developer-tools-now-in-binary-form/29802/1
